Redis 7.2.16 pub/sub tests: wait for no-argument unsubscribe acknowledgments.
The upstream helpers read one reply per supplied name, so no names means no
reply is read. Keep each wire command argument-free and consume its known
memberships before another connection relies on removal. Apply to Firn and
Redis reference runs alike; retire when the pinned suite includes these waits.

Sources:
https://github.com/redis/redis/blob/7.2.16/tests/unit/pubsub.tcl
https://github.com/redis/redis/blob/7.2.16/tests/unit/pubsubshard.tcl

Redis license (COPYING from the same release):
Copyright (c) 2006-2020, Salvatore Sanfilippo
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
    * Neither the name of Redis nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

--- a/tests/unit/pubsub.tcl
+++ b/tests/unit/pubsub.tcl
@@ -84,7 +84,8 @@
     test "PUBLISH/SUBSCRIBE after UNSUBSCRIBE without arguments" {
         set rd1 [redis_deferring_client]
         assert_equal {1 2 3} [subscribe $rd1 {chan1 chan2 chan3}]
-        unsubscribe $rd1
+        $rd1 unsubscribe
+        assert_equal {2 1 0} [consume_subscribe_messages $rd1 unsubscribe {chan1 chan2 chan3}]
         assert_equal 0 [r publish chan1 hello]
         assert_equal 0 [r publish chan2 hello]
         assert_equal 0 [r publish chan3 hello]
@@ -157,7 +158,8 @@
     test "PUBLISH/PSUBSCRIBE after PUNSUBSCRIBE without arguments" {
         set rd1 [redis_deferring_client]
         assert_equal {1 2 3} [psubscribe $rd1 {chan1.* chan2.* chan3.*}]
-        punsubscribe $rd1
+        $rd1 punsubscribe
+        assert_equal {2 1 0} [consume_subscribe_messages $rd1 punsubscribe {chan1.* chan2.* chan3.*}]
         assert_equal 0 [r publish chan1.hi hello]
         assert_equal 0 [r publish chan2.hi hello]
         assert_equal 0 [r publish chan3.hi hello]
@@ -215,8 +217,10 @@
         set patterns [r pubsub numpat]
 
         # clean up clients
-        punsubscribe $rd1
-        punsubscribe $rd2
+        $rd1 punsubscribe
+        assert_equal {1 0} [consume_subscribe_messages $rd1 punsubscribe {foo* bar*}]
+        $rd2 punsubscribe
+        assert_equal {1 0} [consume_subscribe_messages $rd2 punsubscribe {foo* baz*}]
         assert_equal 3 $patterns
         $rd1 close
         $rd2 close
--- a/tests/unit/pubsubshard.tcl
+++ b/tests/unit/pubsubshard.tcl
@@ -45,7 +45,8 @@
         assert_equal {1} [ssubscribe $rd1 {chan1}]
         assert_equal {2} [ssubscribe $rd1 {chan2}]
         assert_equal {3} [ssubscribe $rd1 {chan3}]
-        sunsubscribe $rd1
+        $rd1 sunsubscribe
+        assert_equal {2 1 0} [consume_subscribe_messages $rd1 sunsubscribe {chan1 chan2 chan3}]
         assert_equal 0 [r SPUBLISH chan1 hello]
         assert_equal 0 [r SPUBLISH chan2 hello]
         assert_equal 0 [r SPUBLISH chan3 hello]
